### PR TITLE
registration routes

### PR DESCRIPTION
**Pull request by jacobc2700**
Made routes for:
/save → a database upsert operation to save registration mid-progress
/submit → also a database upsert operation
/get → retrieve registration fields both to repopulate registration info when a user re-enters registration (after saving and closing) and so that we can display a confirmation page at the end

### CODE REVIEW

File: src/services/registration/registration-router.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import { Database } from "../../database";
+import RoleChecker from "../../middleware/role-checker";
+import { Role } from "../auth/auth-models";
+
+const registrationRouter = Router();
+
+// A database upsert operation to save registration mid-progress
+registrationRouter.post(
+    "/save",
+    RoleChecker(["USER"]),
```
Comment (AydanPirani) [inline src/services/registration/registration-router.ts:16]: Should be empty array - anyone can register

File: src/services/registration/registration-router.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    async (req, res, next) => {
+        try {
+            const registrationData = RegistrationValidator.parse(req.body);
+
```
Comment (AydanPirani) [inline src/services/registration/registration-router.ts:20]: add a check in here to see if it's already submitted - if it is, then throw an error

File: src/services/registration/registration-router.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            await Database.REGISTRATION.findOneAndUpdate(
+                { email: registrationData.email }, // only required one
```
Comment (AydanPirani) [inline src/services/registration/registration-router.ts:22]: check based of userId here

File: src/services/registration/registration-router.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+            res.status(StatusCodes.OK).json(registrationData);
+        } catch (error) {
+            next(error);
+        }
+    }
+);
+
+registrationRouter.post(
+    "/submit",
+    RoleChecker(["USER"]),
```
Comment (AydanPirani) [inline src/services/registration/registration-router.ts:39]: open to all

File: src/services/registration/registration-router.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    async (req, res, next) => {
+        try {
+            const registrationData = RegistrationValidator.parse(req.body);
+
+            await Database.REGISTRATION.findOneAndUpdate(
+                { email: registrationData.email }, // only required one
```
Comment (AydanPirani) [inline src/services/registration/registration-router.ts:45]: use userId here

File: src/services/registration/registration-router.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            res.status(StatusCodes.OK).json(registrationData);
+        } catch (error) {
+            next(error);
+        }
+    }
+);
+
+// Retrieve registration fields both to repopulate registration info for a user
+registrationRouter.get(
+    "/get/:email",
```
Comment (AydanPirani) [inline src/services/registration/registration-router.ts:62]: just a /get/
```suggestion
    "/get/",
```

File: src/services/registration/registration-router.ts [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    RoleChecker(["USER"]),
+    async (req, res, next) => {
+        try {
+            const email = req.params.email;
+
+            const registration = await Database.REGISTRATION.findOne({ email });
```
Comment (AydanPirani) [inline src/services/registration/registration-router.ts:68]: find one using {userId: userId}, userId comes from JWT token

File: src/services/registration/registration-schema.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+import { z } from "zod";
+
+// Zod schema for registration
+const RegistrationValidator = z.object({
+    name: z.string().optional(),
+    email: z.string().email(),
+    events: z.array(z.string()).optional(),
+    dietary_restrictions: z.array(z.string()).optional(),
+    points: z.number().min(0).default(0).optional(),
+    complete: z.boolean().default(false),
```
Comment (AydanPirani) [inline src/services/registration/registration-schema.ts:11]:
```suggestion
    hasSubmitted: z.boolean().default(false),
```

File: src/services/registration/registration-schema.ts [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+});
+
+// Mongoose schema for registration
+const RegistrationSchema = new mongoose.Schema({
+    name: { type: String },
+    email: { type: String, required: true, unique: true },
+    events: [{ type: mongoose.Schema.Types.ObjectId, ref: "Event" }],
+    dietary_restrictions: [{ type: String }],
+    points: { type: Number, default: 0 },
+    complete: { type: Boolean, default: false },
```
Comment (AydanPirani) [inline src/services/registration/registration-schema.ts:21]:
```suggestion
    hasCompleted: { type: Boolean, default: false },
```

**Review by AydanPirani — COMMENTED** [review]
few more changes!

**Review by AydanPirani — APPROVED** [review]

File: src/services/registration/registration-router.ts [hunk 10]
```diff
@@ -1,10 +1,88 @@
...
+        );
+
+        return res.status(StatusCodes.OK).json(registrationData);
+    } catch (error) {
+        next(error);
+    }
+});
+
+// Retrieve registration fields both to repopulate registration info for a user
+registrationRouter.get("/get", RoleChecker([]), async (req, res, next) => {
```
Comment (AydanPirani) [inline src/services/registration/registration-router.ts:72]:
```suggestion
registrationRouter.get("/", RoleChecker([]), async (req, res, next) => {
```

**Review by AydanPirani — APPROVED** [review]
1 last super small nit, then we're good to go :)

**Review by divyack2 — APPROVED** [review]
lgtm!

**Review by divyack2 — APPROVED** [review]
lgtm!

**Pull request merged**